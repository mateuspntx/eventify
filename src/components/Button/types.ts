export type BaseButtonProps = {
  title: string;
  variant: 'primary' | 'secondary' | 'outline';
  size: 'default' | 'large';
  isLoading: boolean;
  icon?:
    | import('react').ElementType<import('react-icons').IconBaseProps>
    | JSX.Element;
};

export type ButtonWithIconProps = Omit<BaseButtonProps, 'title' | 'icon'> & {
  title?: never;
  icon:
    | import('react').ElementType<import('react-icons').IconBaseProps>
    | JSX.Element;
};
