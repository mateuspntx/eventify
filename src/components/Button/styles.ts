import styled, { css, DefaultTheme } from 'styled-components';

import { BaseButtonProps } from './types';

type ButtonContainerProps = Pick<BaseButtonProps, 'variant' | 'size'>;

const modifiers = ({ colors }: DefaultTheme) => ({
  variants: {
    primary: css`
      background: ${colors.primary[100]};
    `,
    secondary: css`
      background: ${colors.secondary[100]};
    `,
    outline: css`
      background: none;
      color: ${colors.foreground[100]};
      border: 2px solid ${colors.background[60]};

      :disabled {
        background: none;
        border-color: ${colors.background[40]};
      }
    `,
  },
  sizes: {
    default: css`
      padding: 1rem;
      font-size: 1rem;
      border-radius: 0.5rem;
    `,
    large: css`
      padding: 1.25rem;
      font-size: 1.125rem;
      border-radius: 0.75rem;
    `,
  },
});

export const Container = styled.button<ButtonContainerProps>`
  ${({ theme, variant, size }) => css`
    color: ${theme.colors.base.text};

    > svg + span {
      margin-left: 0.75rem;
    }

    :disabled {
      background: ${theme.colors.background[40]};
    }

    ${modifiers(theme).variants[variant]};
    ${modifiers(theme).sizes[size]};
  `}
`;
