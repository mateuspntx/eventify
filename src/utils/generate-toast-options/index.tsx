import { ComponentProps } from 'react';

import { DefaultTheme as Theme } from 'styled-components';

type Toaster = typeof import('react-hot-toast').Toaster;
type ToastOptions = ComponentProps<Toaster>['toastOptions'];

export const generateToastOptions = ({ colors }: Theme): ToastOptions => ({
  success: {
    iconTheme: {
      primary: colors.success[100],
      secondary: colors.background[100],
    },
  },
  error: {
    iconTheme: {
      primary: colors.error[100],
      secondary: colors.background[100],
    },
  },
});
