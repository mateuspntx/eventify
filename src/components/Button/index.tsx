import { ButtonHTMLAttributes } from 'react';

import * as S from './styles';
import { BaseButtonProps, ButtonWithIconProps } from './types';

export type ButtonProps = (BaseButtonProps | ButtonWithIconProps) &
  ButtonHTMLAttributes<HTMLButtonElement>;

const DEFAULT_CLICK_HANDLER = () => null;

const Button = (props: ButtonProps) => {
  const {
    title,
    onClick,
    variant = 'primary',
    size = 'default',
    isLoading = false,
    type = 'button',
    icon: Icon,
    ...attrs
  } = props;

  const onClickHandler = isLoading ? DEFAULT_CLICK_HANDLER : onClick;
  const iconSize = size === 'default' ? 20 : 24;

  return (
    <S.Container
      type={type}
      variant={variant}
      size={size}
      onClick={onClickHandler}
      {...attrs}
    >
      {Icon ? (
        <>
          {typeof Icon === 'function' ? <Icon size={iconSize} /> : Icon}
          {title && <span>{title}</span>}
        </>
      ) : (
        title
      )}
    </S.Container>
  );
};

Button.defaultProps = {
  isLoading: false,
  variant: 'primary',
  size: 'default',
};

export default Button;
