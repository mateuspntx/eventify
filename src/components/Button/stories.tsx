import { Story, Meta } from '@storybook/react/types-6-0';
import { FiGrid } from 'react-icons/fi';

import Button, { ButtonProps } from '.';

export default {
  title: 'Button',
  component: Button,
} as Meta;

type TypedStory = Story<ButtonProps>;

const Template: TypedStory = (args) => <Button {...args} />;

export const Primary: TypedStory = Template.bind({});
Primary.args = {
  title: 'Primary Button',
};

export const Secondary: TypedStory = Template.bind({});
Secondary.args = {
  title: 'Secondary Button',
  variant: 'secondary',
};

export const Outline: TypedStory = Template.bind({});
Outline.args = {
  title: 'Outline Button',
  variant: 'outline',
};

export const WithIcon: TypedStory = Template.bind({});
WithIcon.args = {
  title: 'With Icon Button',
  variant: 'primary',
  icon: FiGrid,
};

export const OnlyIcon: TypedStory = Template.bind({});
OnlyIcon.args = {
  icon: FiGrid,
};
