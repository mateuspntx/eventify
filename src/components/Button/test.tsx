import { screen } from '@testing-library/react';
import { FiSend } from 'react-icons/fi';

import { renderWithTheme } from '@utils/tests/utils';

import Button from '.';

describe('<Button />', () => {
  it('should render the button', () => {
    const { container } = renderWithTheme(<Button title="Submit" />);

    const renderedButton = screen.getByRole('button', { name: /submit/i });

    expect(renderedButton).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the button with icon', () => {
    const { container } = renderWithTheme(
      <Button title="Submit" icon={<FiSend data-testid="icon-test" />} />,
    );

    const renderedButton = screen.getByRole('button', { name: /submit/i });
    const renderedIcon = screen.getByTestId('icon-test');

    expect(renderedButton).toBeInTheDocument();
    expect(renderedButton.childNodes).toHaveLength(2);

    expect(renderedIcon).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the button with only the icon', () => {
    const { container } = renderWithTheme(
      <Button icon={<FiSend data-testid="icon-test" />} />,
    );

    const renderedButton = screen.getByRole('button');
    const renderedIcon = screen.getByTestId('icon-test');

    expect(renderedButton).toBeInTheDocument();
    expect(renderedButton.childNodes).toHaveLength(1);

    expect(renderedIcon).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
