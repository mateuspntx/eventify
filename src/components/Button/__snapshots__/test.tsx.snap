// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button 1`] = `
.c0 {
  color: var(--colors-text-base);
  background: var(--colors-primary-100);
  padding: 1rem;
  font-size: 1rem;
  border-radius: 0.5rem;
}

.c0 > svg + span {
  margin-left: 0.75rem;
}

.c0:disabled {
  background: var(--colors-background-40);
}

<button
  class="c0"
  type="button"
>
  Submit
</button>
`;

exports[`<Button /> should render the button with icon 1`] = `
.c0 {
  color: var(--colors-text-base);
  background: var(--colors-primary-100);
  padding: 1rem;
  font-size: 1rem;
  border-radius: 0.5rem;
}

.c0 > svg + span {
  margin-left: 0.75rem;
}

.c0:disabled {
  background: var(--colors-background-40);
}

<button
  class="c0"
  type="button"
>
  <svg
    data-testid="icon-test"
    fill="none"
    height="1em"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <line
      x1="22"
      x2="11"
      y1="2"
      y2="13"
    />
    <polygon
      points="22 2 15 22 11 13 2 9 22 2"
    />
  </svg>
  <span>
    Submit
  </span>
</button>
`;

exports[`<Button /> should render the button with only the icon 1`] = `
.c0 {
  color: var(--colors-text-base);
  background: var(--colors-primary-100);
  padding: 1rem;
  font-size: 1rem;
  border-radius: 0.5rem;
}

.c0 > svg + span {
  margin-left: 0.75rem;
}

.c0:disabled {
  background: var(--colors-background-40);
}

<button
  class="c0"
  type="button"
>
  <svg
    data-testid="icon-test"
    fill="none"
    height="1em"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <line
      x1="22"
      x2="11"
      y1="2"
      y2="13"
    />
    <polygon
      points="22 2 15 22 11 13 2 9 22 2"
    />
  </svg>
</button>
`;
